File: examples/scala/mega_flow/Storage.scala
package mega

// Class hierarchy — abstract / trait / case class, inheritance,
// override + super — all preserving taint on the way to the sink.
trait Auditable {
  def audit(): Unit = ()
}

abstract class BaseRepository(val data: App.Envelope) extends Auditable {
  def cmd: String = data.cmd
  def run(): String
}

class Repository(data: App.Envelope) extends BaseRepository(data) {
  override def run(): String = {
    val c = cmd
    Executor.execute(c)
  }
}

class AuditedRepository(data: App.Envelope) extends Repository(data) {
  override def run(): String = {
    audit()
    // super-call preserves taint across the inheritance chain.
    super.run()
  }
}

object Repository {
  def wrap(data: App.Envelope): AuditedRepository = new AuditedRepository(data)
}

object Storage {
  def persist(envelope: App.Envelope): String = Repository.wrap(envelope).run()
}

File: examples/dart/mega_flow/storage.dart
// Class hierarchy — abstract base, mixins, inheritance, factory
// constructors — all preserving taint on the way to the sink.
import 'app.dart';
import 'executor.dart';

mixin Auditable {
  void audit() {}
}

abstract class BaseRepository {
  final Envelope data;
  BaseRepository(this.data);

  // Getter exposes the tainted cmd field.
  String get cmd => data.cmd;

  String run();
}

class Repository extends BaseRepository with Auditable {
  Repository(super.data);

  factory Repository.wrap(Envelope data) => Repository(data);

  @override
  String run() {
    final c = cmd;
    return execute(c);
  }
}

class AuditedRepository extends Repository {
  AuditedRepository(super.data);

  @override
  String run() {
    audit();
    // super-call preserves taint across the inheritance chain.
    return super.run();
  }
}

String persist(Envelope envelope) {
  final repo = AuditedRepository(envelope);
  return repo.run();
}
